math/complex_numbers.cpp

```diff
@@ -0,0 +1,190 @@
+     * @return result current number subtract other number
+     */
+    Complex operator-(const Complex &other) {
+        Complex result(this->re - other.re, this->im - other.im);
+        return result;
+    }
+
+    /**
+     * Operator overload to be able to multiple two complex numbers.
+     * @param other The other number to multiply the current number to.
+     * @return result current number times other number.
+     */
+    Complex operator*(const Complex &other) {
+        Complex result(this->re * other.re - this->im * other.im,
+                       this->re * other.im + this->im * other.re);
+        return result;
+    }
+
+    /**
+     * Operator overload of the BITWISE NOT which gives us the conjugate of our
+     * complex number. NOTE: This is overloading the BITWISE operator but its
+     * not a BITWISE operation in this definition.
+     * @return result The conjugate of our complex number.
+     */
+    Complex operator~() const {
+        Complex result(this->re, -(this->im));
+        return result;
+    }
+
+    /**
+     * Operator overload to be able to divide two complex numbers. This function
+     * would throw an exception if the other number is zero.
+     * @param other The other number we divide our number by.
+     * @return result Current number divided by other number.
+     */
+    Complex operator/(const Complex &other) {
+        Complex result = *this * ~other;
+        double denominator = other.abs() * other.abs();
+        if (denominator != 0) {
+            result = Complex(result.real() / denominator,
+                             result.imag() / denominator);
+            return result;
+        } else {
+            throw std::invalid_argument("Undefined Value");
+        }
+    }
+};
+
+/**
+ * Logical Equal overload for our Complex class.
+ * @param a Left hand side of our expression
+ * @param b Right hand side of our expression
+ * @return 'True' If real and imaginary parts of a and b are same
+ * @return 'False' Otherwise.
+ */
+bool operator==(const Complex &a, const Complex &b) {
+    double del_real = a.real() - b.real();
+    double del_imag = a.imag() - b.imag();
+    return ((del_real <= 1e-15 && del_real >= -1e-15) &&
+            (del_imag <= 1e-15 && del_imag >= -1e-15));
+}
+
+/**
+ * Overloaded insersion operator to accommodate the printing of our complex
+ * number in their standard form.
+ * @param os The console stream
+ * @param num The complex number.
+ */
+std::ostream &operator<<(std::ostream &os, const Complex &num) {
+    os << num.real();
+    if (num.imag() < 0) {
+        os << " - " << -num.imag();
+    } else {
+        os << " + " << num.imag();
+    }
+    os << "i";
+    return os;
+}
+
+/**
+ * Tests Function
+ */
+void tests() {
+    Complex num1(1, 1), num2(1, 1);
+    // Test for addition
+    assert(((void)"1 + 1i + 1 + 1i is equal to 2 + 2i but the addition doesn't "
+                  "add up \n",
+            (num1 + num2) == Complex(2, 2)));
```

Comment:
I think the correct way to check would be to compare with the results from `std::complex` class.

Reply:
All tests work except for the one that checks for division.

Reply:
I changed my implementation to coincide with the standard libraries.